File: archive/code/src/sync/discovery/jsonl.ts
import { existsSync, type Dirent } from "node:fs";
import { open, readdir, stat } from "node:fs/promises";
import { join } from "node:path";

import { findNearestAlmanacDir } from "../../paths.js";
import type { SessionCandidate, SweepApp } from "./types.js";

export async function collectJsonl(root: string): Promise<string[]> {
  if (!existsSync(root)) return [];
  const out: string[] = [];
  await collectJsonlInto(root, out);
  return out;
}

export async function readFirstLines(file: string, maxLines: number): Promise<string[]> {
  const handle = await open(file, "r").catch(() => null);
  if (handle === null) return [];
  try {
    const buffer = Buffer.alloc(64 * 1024);
    const { bytesRead } = await handle.read(buffer, 0, buffer.length, 0);
    return buffer.subarray(0, bytesRead).toString("utf8").split(/\r?\n/).slice(0, maxLines);
  } finally {
    await handle.close();
  }
}

export async function candidateFromMeta(
  app: SweepApp,
  transcriptPath: string,
  sessionId: string,
  cwd: string,
): Promise<SessionCandidate | null> {
  const repoRoot = findNearestAlmanacDir(cwd);
  if (repoRoot === null) return null;
  try {
    const st = await stat(transcriptPath);
    if (!st.isFile()) return null;
    return {
      app,
      sessionId,
      transcriptPath,
      cwd,
      repoRoot,
      mtimeMs: st.mtimeMs,
      sizeBytes: st.size,
    };
  } catch {
    return null;
  }
}

export function parseJsonObject(line: string): Record<string, unknown> | null {
  if (line.trim().length === 0) return null;
  try {
    const parsed = JSON.parse(line) as unknown;
    return parsed !== null && typeof parsed === "object" && !Array.isArray(parsed)
      ? parsed as Record<string, unknown>
      : null;
  } catch {
    return null;
  }
}

export function objectField(
  obj: Record<string, unknown>,
  key: string,
): Record<string, unknown> | undefined {
  const value = obj[key];
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : undefined;
}

export function stringField(obj: Record<string, unknown>, key: string): string | undefined {
  const value = obj[key];
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

async function collectJsonlInto(dir: string, out: string[]): Promise<void> {
  let entries: Dirent[];
  try {
    entries = await readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      await collectJsonlInto(full, out);
    } else if (entry.isFile() && entry.name.endsWith(".jsonl")) {
      out.push(full);
    }
  }
}

File: src/codealmanac/integrations/sources/transcripts/jsonl.py
import json
from datetime import UTC, datetime
from pathlib import Path

from codealmanac.core.paths import normalize_path
from codealmanac.services.sources.models import TranscriptApp, TranscriptCandidate


def collect_jsonl(root: Path) -> tuple[Path, ...]:
    if not root.is_dir():
        return ()
    return tuple(sorted(path for path in root.rglob("*.jsonl") if path.is_file()))


def read_first_lines(path: Path, max_lines: int) -> tuple[str, ...]:
    try:
        with path.open("r", encoding="utf-8") as file:
            lines = []
            for _ in range(max_lines):
                line = file.readline()
                if line == "":
                    break
                lines.append(line.rstrip("\r\n"))
            return tuple(lines)
    except OSError:
        return ()


def parse_json_object(line: str) -> dict[str, object] | None:
    if not line.strip():
        return None
    try:
        parsed = json.loads(line)
    except ValueError:
        return None
    if isinstance(parsed, dict):
        return parsed
    return None


def object_field(
    value: dict[str, object],
    key: str,
) -> dict[str, object] | None:
    field = value.get(key)
    if isinstance(field, dict):
        return field
    return None


def string_field(value: dict[str, object], key: str) -> str | None:
    field = value.get(key)
    if isinstance(field, str) and field:
        return field
    return None


def candidate_from_meta(
    app: TranscriptApp,
    transcript_path: Path,
    session_id: str,
    cwd: str,
) -> TranscriptCandidate | None:
    transcript_cwd = normalize_path(Path(cwd))
    try:
        stat = transcript_path.stat()
    except OSError:
        return None
    if not transcript_path.is_file():
        return None
    return TranscriptCandidate(
        app=app,
        session_id=session_id,
        transcript_path=normalize_path(transcript_path),
        cwd=transcript_cwd,
        modified_at=datetime.fromtimestamp(stat.st_mtime, UTC),
        size_bytes=stat.st_size,
    )
